scripts/ECOINVENT_MIGRATION_SCRIPT_PACKAGE_20230614/Ecoinvent_uuid_migration_jython_script.py

```diff
@@ -0,0 +1,130 @@
+  'H:Transportation and storage',
+  'I:Accommodation and food service activities',
+  'J:Information and communication',
+  'M:Professional, scientific and technical activities',
+  'N:Administrative and support service activities',
+  'S:Other service activities']
+
+
+def filter_processes():
+  """This function will split the foreground and background processes in the database. For this function to work none, of the foreground processes should be saved inside any of the original folders of the ecoinvent database. This function returns two lists, the first contains all foreground processes and the second all background processes."""
+  
+  foreground_process_category_list = []
+  background_process_category_list = []
+  for process in db.getAll(Process):
+    category = process.category
+    cat_list = [category.name]
+    
+    def loop(category_obj):
+      if category_obj != None:
+        if category_obj.category != None:
+          cat_list.append(category_obj.category.name)
+        loop(category_obj.category) 
+     
+    loop(category)
+    cat_list.reverse()
+    cat_path = '/'.join(cat_list)
+    
+    if any(elem in cat_path for elem in ECOINVENT_PARENT_CATEGORY_LIST):
+      background_process_category_list.append(process)
+    else:
+      foreground_process_category_list.append(process)
+    
+  return foreground_process_category_list, background_process_category_list
+     
+
+def migrate_providers_of_all_foreground_processes(mapping_dict):
+  """This function exchanges all the providers of all the foreground processes inside the database if an analog exists in the mapping dictionary which will be derived from an external json file."""
+  
+  foreground_process_list = filter_processes()[0]
+  process_dao = ProcessDao(db)
+  for foreground_process in foreground_process_list:
+    for exchange in foreground_process.exchanges:
+      if exchange.defaultProviderId != 0:
+        provider = db.get(Process, exchange.defaultProviderId)
+        if provider.refId in mapping_dict:
+          new_provider = db.get(Process, mapping_dict[provider.refId][0])
+          exchange.defaultProviderId = new_provider.id
+    process_dao.update(foreground_process)
+    print('Updated providers in foreground process "%s"' % foreground_process.refId)
+  for process in db.getAll(Process):
+    if process.refId in mapping_dict:
+      old_ref_id = process.refId
+      process_dao.delete(process)
+      print('Deleted oudated Ecoinvent3.8 process "%s"' % old_ref_id)
+      
+def update_product_system():
+  """This function reevaluates the existing product systems inside the database."""
+  
+  product_system_dao = ProductSystemDao(db)
+  for product_system in db.getAll(ProductSystem):
+    p = product_system.referenceProcess
+    config = LinkingConfig().providerLinking(ProviderLinking.PREFER_DEFAULTS).preferredType(ProcessType.UNIT_PROCESS)
+    new_product_system = ProductSystemBuilder(MatrixCache.createLazy(db), config).build(p)
```

Comment:
Such piece of code can be written that way:
```suggestion
config = LinkingConfig()\
            .providerLinking(ProviderLinking.PREFER_DEFAULTS)\
            .preferredType(ProcessType.UNIT_PROCESS)
new_product_system = ProductSystemBuilder(MatrixCache.createLazy(db), config)\
            .build(p)
```